updater/internal/file/file.go

```diff
@@ -43,21 +43,42 @@ func CopyFile(logger *zap.Logger, pathIn, pathOut string, overwrite bool) error
 
 	pathOutClean := filepath.Clean(pathOut)
 	fileMode := fs.FileMode(0600)
-
 	flags := os.O_CREATE | os.O_WRONLY
 	if overwrite {
 		// If we are OK to overwrite, we will truncate the file on open
 		flags |= os.O_TRUNC
 
-		// Save old file's permissions and delete it first if it exists
-		fileOutInfo, _ := os.Stat(pathOutClean)
-		if fileOutInfo != nil {
-			fileMode = fileOutInfo.Mode()
+		// Try to save old file's permissions
+		outFileInfo, _ := os.Stat(pathOutClean)
+		if outFileInfo != nil {
+			fileMode = outFileInfo.Mode()
+		} else if useInFilePermBackup {
+			// Use the new file's permissions as a backup and don't fail on error (best chance for rollback)
+			inFileInfo, err := inFile.Stat()
+			if err != nil {
+				logger.Error("failed to retrieve fileinfo for input file", zap.Error(err))
+			}
+			if inFileInfo != nil {
+				fileMode = inFileInfo.Mode()
+			}
```

Comment:
This check seems odd. If `err != nil` likely `inFileInfo == nil` These two cases are mutually exclusive so only one can happen.

It might read better as a switch statement:

```go
switch {
  case err != nil:
    logger.Error("")
  case inFileInfo != nil:
    fileMode = inFileInfo.Mode()
}
```

Reply:
@cpheps i just wasn't sure if something weird could happen like an error occurred but there was still file info. but yeah, it probably doesn't make sense. i can change this.